source/extensions/common/wasm/wasm.cc

```diff
@@ -934,65 +952,75 @@ Wasm::Wasm(absl::string_view vm, absl::string_view id, absl::string_view initial
-    _REGISTER_PROXY(addRequestHeader);
-    _REGISTER_PROXY(replaceRequestHeader);
-    _REGISTER_PROXY(removeRequestHeader);
-    _REGISTER_PROXY(getRequestHeaderPairs);
-
-    _REGISTER_PROXY(getRequestTrailer);
-    _REGISTER_PROXY(addRequestTrailer);
-    _REGISTER_PROXY(replaceRequestTrailer);
-    _REGISTER_PROXY(removeRequestTrailer);
-    _REGISTER_PROXY(getRequestTrailerPairs);
-
-    _REGISTER_PROXY(getResponseHeader);
-    _REGISTER_PROXY(addResponseHeader);
-    _REGISTER_PROXY(replaceResponseHeader);
-    _REGISTER_PROXY(removeResponseHeader);
-    _REGISTER_PROXY(getResponseHeaderPairs);
-
-    _REGISTER_PROXY(getResponseTrailer);
-    _REGISTER_PROXY(addResponseTrailer);
-    _REGISTER_PROXY(replaceResponseTrailer);
-    _REGISTER_PROXY(removeResponseTrailer);
-    _REGISTER_PROXY(getResponseTrailerPairs);
-
-    _REGISTER_PROXY(getRequestBodyBufferBytes);
-    _REGISTER_PROXY(getResponseBodyBufferBytes);
-
-    _REGISTER_PROXY(httpCall);
-
-    _REGISTER_PROXY(setTickPeriodMilliseconds);
+  // Calls with the "_proxy_" prefix.
+#define _REGISTER_PROXY(_fn)                                                                       \
+  registerCallback(wasm_vm_.get(), "envoy", "_proxy_" #_fn, &_fn##Handler);
+  _REGISTER_PROXY(log);
+
+  _REGISTER_PROXY(getRequestStreamInfoProtocol);
+  _REGISTER_PROXY(getResponseStreamInfoProtocol);
+
+  _REGISTER_PROXY(getRequestMetadata);
+  _REGISTER_PROXY(setRequestMetadata);
+  _REGISTER_PROXY(getRequestMetadataPairs);
+  _REGISTER_PROXY(getResponseMetadata);
+  _REGISTER_PROXY(setResponseMetadata);
+  _REGISTER_PROXY(getResponseMetadataPairs);
+
+  _REGISTER_PROXY(continueRequest);
+  _REGISTER_PROXY(continueResponse);
+
+  _REGISTER_PROXY(getSharedData);
+  _REGISTER_PROXY(setSharedData);
+
+  _REGISTER_PROXY(getRequestHeader);
+  _REGISTER_PROXY(addRequestHeader);
+  _REGISTER_PROXY(replaceRequestHeader);
+  _REGISTER_PROXY(removeRequestHeader);
+  _REGISTER_PROXY(getRequestHeaderPairs);
+
+  _REGISTER_PROXY(getRequestTrailer);
+  _REGISTER_PROXY(addRequestTrailer);
+  _REGISTER_PROXY(replaceRequestTrailer);
+  _REGISTER_PROXY(removeRequestTrailer);
+  _REGISTER_PROXY(getRequestTrailerPairs);
+
+  _REGISTER_PROXY(getResponseHeader);
+  _REGISTER_PROXY(addResponseHeader);
+  _REGISTER_PROXY(replaceResponseHeader);
+  _REGISTER_PROXY(removeResponseHeader);
+  _REGISTER_PROXY(getResponseHeaderPairs);
+
+  _REGISTER_PROXY(getResponseTrailer);
+  _REGISTER_PROXY(addResponseTrailer);
+  _REGISTER_PROXY(replaceResponseTrailer);
+  _REGISTER_PROXY(removeResponseTrailer);
+  _REGISTER_PROXY(getResponseTrailerPairs);
+
+  _REGISTER_PROXY(getRequestBodyBufferBytes);
+  _REGISTER_PROXY(getResponseBodyBufferBytes);
+
+  _REGISTER_PROXY(httpCall);
+
+  _REGISTER_PROXY(setTickPeriodMilliseconds);
 #undef _REGISTER_PROXY
+}
+
+void Wasm::establishEnvironment() {
+  if (is_emscripten_) {
+    wasm_vm_->makeModule("global");
+    emscripten_NaN_ = makeGlobal(wasm_vm_.get(), "global", "NaN", std::nan("0"));
```

Comment:
"global", "NaN" and "Infinity" are core concept. Could you use const variable and aggregate them in single place?